main/Main.hs

```diff
@@ -23,59 +26,95 @@ import Nixfmt
 import System.IO.Atomic (withOutputFile)
 import System.IO.Utf8 (readFileUtf8, withUtf8StdHandles)
 
-type Result = Either ParseErrorBundle ()
+type Result = Either String ()
 
 data Nixfmt = Nixfmt
     { files :: [FilePath]
     , width :: Int
+    , check :: Bool
+    , quiet :: Bool
     } deriving (Show, Data, Typeable)
 
 options :: Nixfmt
 options = Nixfmt
     { files = [] &= args &= typ "FILES"
-    , width = 80 &= help "Maximum width of the formatted file"
-    } &= help "Format Nix source code"
-
-formatStdio :: Int -> IO Result
-formatStdio w = do
-    contents <- TextIO.getContents
-    mapM TextIO.putStr $ format w "<stdin>" contents
-
-formatFile :: Int -> FilePath -> IO Result
-formatFile w path = do
-    contents <- readFileUtf8 path
-    mapM atomicWriteFile $ format w path contents
+    , width = 80 &= help "Maximum width in characters"
+    , check = False &= help "Check whether files are formatted"
+    , quiet = False &= help "Do not report errors"
+    } &= summary "Format Nix source code"
+
+format' :: Int -> FilePath -> Text -> Either String Text
+format' w path = first errorBundlePretty . format w path
+
+data Target = Target (IO Text) FilePath (Text -> IO ())
+
+formatTarget :: Int -> Target -> IO Result
+formatTarget w (Target doRead path doWrite)
+    = doRead >>= mapM doWrite . format' w path
```

Comment:
It’s very uncool when the data moves in two different directions within a single line. Try:
```suggestion
    = doRead >>= format' w path & mapM doWrite
```
(I’m not 100% if the precedences are right here, but, I think, they are.)

Reply:
This did not typecheck, even after parenthesizing, so I used do-notation in the end.